### PR TITLE
backport: merge bitcoin#19326, #20464, #23451, #20852, #20355, #22050, #23115, #23409, #18468, #23413, partial bitcoin#21560, #23438 (auxiliary backports: part 11)

### PR DESCRIPTION
**Pull request by kwvg**
## Additional information

* [bitcoin#19326](https://github.com/bitcoin/bitcoin/pull/19326) was partially backported as part of [dash#4164](https://github.com/dashpay/dash/pull/4164) to the exclusion of 2a2182c387f607cd8284f33890bd285a81077b7f, 77c507358bda9bd6c496f33e0f4418c0603bb08d.
  The excluded commits have been backported, marking the pull request as fully merged.
* [bitcoin#18468](https://github.com/bitcoin/bitcoin/pull/19326) was partially backported as part of [dash#4036](https://github.com/dashpay/dash/pull/4036) to the exclusion of 2676aeadfa0e43dcaaccc4720623cdfe0beed528.
   The excluded commit has been backported, marking the pull request as fully merged.
* [bitcoin#21560](https://github.com/bitcoin/bitcoin/pull/21560) does not include any TorV3 nodes.
* When backporting [bitcoin#20464](https://github.com/bitcoin/bitcoin/pull/20464), `llmq/dkgsessionhandler.cpp` needs a `reinterpret_cast` because `CHashWriter` accepts a `const char*` while `CDataStream`'s `data()` returns an `unsigned const char*`. 

  Using `reinterpret_cast` is acceptable because `CHashWriter::write()` simply casts the `const char*` to `unsigned const char*`, reversing the previous cast ([source](https://github.com/dashpay/dash/blob/38e70430b943021e6d4d6863ee53d7cb63c4fd16/src/hash.h#L202-L204)), making this exercise solely for the purpose of fitting the pointer through the argument.

  A C-style cast wasn't used because the linter took objection to it ([build](https://gitlab.com/dashpay/dash/-/jobs/5051487709#L105)) and only it, but seemed to be fine with every other C-style cast.
* Dependency for https://github.com/dashpay/dash/pull/5575
  * [bitcoin#27479](https://github.com/bitcoin/bitcoin/pull/27479), the primary motivation for [dash#5575](https://github.com/dashpay/dash/pull/5575) uses features added to `Span` (such as the removal of `MakeSpan` and the introduction of `BytePtr`)
  * Another backport that was primarily required as [bitcoin#22570](https://github.com/bitcoin/bitcoin/pull/22570), which does away with the serialization code for `CSubNet`, which utilizes `Span` but would be incompatible with [bitcoin#23413](https://github.com/bitcoin/bitcoin/pull/23413), at least on macOS ([build](https://gitlab.com/dashpay/dash/-/jobs/5051487719#L1769))
    * This in turn pulled in a bunch of networking-related backports and only the minimum set required to backport [bitcoin#22570](https://github.com/bitcoin/bitcoin/pull/22570) has been included. 
    
        This has the effect of making [bitcoin#17428](https://github.com/bitcoin/bitcoin/pull/17428) vestigial as the code that actually integrates it with existing logic has been excluded from the backport as it relies on larger networking backports which are far beyond the scope of this PR.
  * This in turn required backporting the contributions made before it, which revealed merge conflicts, which then needed to be added into the fold of backports included in the pull request in order to ensure the least amount of conflicts.
 
## Warnings

* The greater adoption of `Span`, removal of `MakeSpan` and introduction of `BytePtr` involved changes made to taproot and segregated witness code. This pull request may complicate Taproot backporting should that ever be on the roadmap.

* `BytePtr` is later renamed to `AsBytePtr` due to a naming conflict with the iOS SDK (see discussion for [bitcoin#24989](https://github.com/bitcoin/bitcoin/pull/24989)) but in the interest of minimizing the set of changes needed to merge in [dash#5575](https://github.com/dashpay/dash/pull/5575) and the fact that Dash does not target i(Pad)OS, this is considered acceptable for now.

### CODE REVIEW

File: src/test/net_tests.cpp [hunk 1]
```diff
@@ -281,16 +281,6 @@ BOOST_AUTO_TEST_CASE(cnetaddr_basic)
     BOOST_CHECK(!addr.IsBindAny());
     BOOST_CHECK_EQUAL(addr.ToString(), link_local);
 
-    // TORv2
```
Comment (UdjinM6) [inline src/test/net_tests.cpp:265]: 22050: missing
```
    // TORv2, no longer supported
    BOOST_CHECK(!addr.SetSpecial("6hzph5hv6337r6p2.onion"));
```

File: doc/files.md [hunk 2]
```diff
@@ -116,6 +115,7 @@ These subdirectories and files are no longer used by Dash Core:
 
 Path           | Description | Repository notes
 ---------------|-------------|-----------------
+`banlist.dat`  | Stores the addresses/subnets of banned nodes; superseded by `banlist.json` in 22.0 and completely ignored in 23.0 | [PR #20966](https://github.com/bitcoin/bitcoin/pull/20966), [PR #22570](https://github.com/bitcoin/bitcoin/pull/22570)
```
Comment (UdjinM6) [inline doc/files.md:118]: 22570: should be dashified

File: src/addrdb.cpp [hunk 3]
```diff
@@ -198,17 +198,16 @@ bool CBanDB::Write(const banmap_t& banSet)
     return false;
 }
 
-bool CBanDB::Read(banmap_t& banSet, bool& dirty)
+bool CBanDB::Read(banmap_t& banSet)
 {
-    // If the JSON banlist does not exist, then try to read the non-upgraded banlist.dat.
+    if (fs::exists(m_banlist_dat)) {
+        LogPrintf("banlist.dat ignored because it can only be read by " PACKAGE_NAME " version 22.x. Remove %s to silence this warning.\n", m_banlist_dat);
```
Comment (UdjinM6) [inline src/addrdb.cpp:204]: 22570: should be dashified

File: src/test/net_tests.cpp (hunk 1 above)
Comment (kwvg) [reply src/test/net_tests.cpp:265]: Resolved in latest push.

File: doc/files.md (hunk 2 above)
Comment (kwvg) [reply doc/files.md]: Resolved in latest push.

File: src/addrdb.cpp (hunk 3 above)
Comment (kwvg) [reply src/addrdb.cpp]: Resolved in latest push.

File: src/hash.h [hunk 4]
```diff
@@ -84,25 +84,19 @@ class CHash160 {
...
+    CHash256().Write(MakeUCharSpan(in1)).Finalize(result);
     return result;
 }
 
 /** Compute the 256-bit hash of the concatenation of two objects. */
 template<typename T1, typename T2>
-inline uint256 Hash(const T1 p1begin, const T1 p1end,
-                    const T2 p2begin, const T2 p2end) {
-    static const unsigned char pblank[1] = {};
+inline uint256 Hash(const T1& in1, const T2& in2) {
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/hash.h:97]: Functions are unused and have been removed. Resolved in latest push.

**Review by knst — CHANGES_REQUESTED** [review]
as discussed in slack, please, use Span class here instead std::vector for scriptPubKey's + Hash160

Here's possible implementation: https://github.com/knst/dash/commit/a534b5f9edef94d63bbc67071787419edc8c263a

**Comment by PastaPastaPasta** [comment]
> as discussed in slack, please, use Span class here instead std::vector for scriptPubKey's + Hash160
> 
> Here's possible implementation: [knst@a534b5f](https://github.com/knst/dash/commit/a534b5f9edef94d63bbc67071787419edc8c263a)

This isn't needed? The purpose of using span is you can implicitly construct one from basically anything; seems fine to me to keep it as is

**Comment by knst** [comment]
> This isn't needed? The purpose of using span is you can implicitly construct one from basically anything; seems fine to me to keep it as is

These 2 commits are extra context for you @PastaPastaPasta 
https://github.com/knst/dash/commit/5d91e7ecc4c1b00822ff90c2a8f99242d7d5f9df
https://github.com/knst/dash/commit/7d80ffbbb575160aca4323fd64836d05b3d06498

so, for this code will cause compilation error with new Span implementation (after all backports from this batch):
```cpp
uint160 hashBytes(Hash160(prevout.scriptPubKey.begin()+1, prevout.scriptPubKey.end()-1));
```
with new Span implementation compiler will throw an error

kwvg's solution:
```diff
-uint160 hashBytes(Hash160(prevout.scriptPubKey.begin()+1, prevout.scriptPubKey.end()-1));
+uint160 hashBytes(Hash160(std::vector<unsigned char>(prevout.scriptPubKey.begin()+1, prevout.scriptPubKey.end()-1)));
```
my patch:
```diff
-uint160 hashBytes(Hash160(prevout.scriptPubKey.begin()+1, prevout.scriptPubKey.end()-1));
+uint160 hashBytes{Hash160(Span{prevout.scriptPubKey.data()+1, prevout.scriptPubKey.size() - 2})};
```

my solution uses Span instead std::vector and it helps to avoid extra memory allocation, because Span do now copy data, do not own it, it is just container with pointer to data + its length (same as std::span and std::string_view), but std::vector<...> will cause memory allocation + memory copy + memory deallocation when its data no need more.

**Comment by PastaPastaPasta** [comment]
@knst thanks for the context! I didn't see that current code was allocating a vector.

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by knst — APPROVED** [review]
LGTM

**Pull request merged**